server/session.go

```diff
@@ -466,21 +504,87 @@ func (p *Plugin) removeSession(us *session) error {
 		return nil
 	}
 
-	state, err := p.lockCallReturnState(us.channelID)
+	sessionsCount, err := p.store.GetCallSessionsCount(us.callID, db.GetCallSessionOpts{})
 	if err != nil {
-		return fmt.Errorf("failed to lock call: %w", err)
+		p.LogError("failed to get call sessions count", "callID", us.callID, "err", err.Error())
 	}
-	defer p.unlockCall(us.channelID)
 
-	p.LogDebug("removing session from state", "userID", us.userID, "connID", us.connID, "originalConnID", us.originalConnID)
+	removeSessionFromCall := func(state *callState) {
+		p.LogDebug("removing session from state", "userID", us.userID, "connID", us.connID, "originalConnID", us.originalConnID)
+
+		p.mut.Lock()
+		delete(p.sessions, us.connID)
+		p.mut.Unlock()
+
+		if err := p.removeUserSession(state, us.userID, us.originalConnID, us.connID, us.channelID); err != nil {
+			p.LogError("failed to remove user session ", "originalConnID", us.originalConnID, "err", err.Error())
+		}
+	}
 
 	p.mut.Lock()
-	delete(p.sessions, us.connID)
+	batcher := p.removeSessionsBatchers[us.channelID]
+	shouldBatch := batcher != nil || sessionsCount >= minMembersCountForBatching
```

Comment:
So only batch when we're past the `minMembersCountForBatching` threshold, but I guess there's no way to tell if you will actually need the batching ahead of time, right?  (I can't think of a way, other than adding an 'end call' button... But even then it wouldn't detect the mass leaving just before.) Just thinking out loud.

Reply:
The leaving case is actually easier since we know how many sessions we have connected and can assume eventually they will disconnect roughly at the same time as that's just the nature of large (planned) calls, likely within a tight schedule.

When joining the call it's harder because we have no idea, see the comment I left:

https://github.com/mattermost/mattermost-plugin-calls/blob/c2623a7b86f0aeeda629bf969613b6e68b7ce82e/server/websocket.go#L892-L895